### DISCUSSION TITLE
[Bug]: Warning for Cruisecontrol reported: Skip generating metric sample for broker 1 because all broker metrics are missing.

### DISCUSSION THREAD

**Discussion by nsettipa** [strimzi/strimzi-kafka-operator#10626] [General]
### Bug Description

we are using kafka for messaging and events generation. Producer and consumer traffic ongoing and once in a while below warning reported for metrics collection. 
Skip generating metric sample for broker 1 because all broker metrics are missing. 
This warning reported for cruisecontrol when kubectl logs captured and analyzed the logs. 
This warning reported while traffic running. 

Does this have any impact on metrics collection or brokers performance as such. This issue observed on latest cruise-control and kafka versions. 

### Expected behavior

we would like to know if this warning can be ignored or causing functional impact.

### Strimzi version

latest strimzi version

### Kubernetes version

1.29

### Installation method

installed pods through k apply -f kafka yaml where cruisecontrol also activated

### Infrastructure

K8s cluster on linux

**Comment by scholzj** [comment]
I guess a start would be to explain what exactly you are doing, share your configurations, full logs, explain what exactly you see and where etc.

**Comment by nsettipa** [comment]
We are simulating the traffic with 3 kafka broker Pods. One cruise control Pod running which monitors the 3 kafka broker pods. Producer and consumer traffic running. during traffic handling time, below warning error reported:

"Skip generating metric sample for broker 2 because all broker metrics are missing.", "metadata": {"container_name": "cruise-control", "pod_name": "cruise-control-5489c5488f-nld6q"}, "extra_data": {"file": "SamplingUtils.java", "line": "295", "mbkf_cluster_name"

**Comment by nsettipa** [comment]
"warning", "service_id": "Skip generating metric sample for broker 2 because all broker metrics are missing.", "metadata": {"container_name": "cruise-control", "pod_name": "cruise-control-5489c5488f-nld6q"}, "extra_data": {"file": "SamplingUtils.java", "line": "295", "mbkf_cluster_name"}}

**Comment by nsettipa** [comment]
Please share your analysis. Is there possibility to disable the warning.

**Comment by fvaleri** [comment]
Hi @nsettipa, [here](https://github.com/linkedin/cruise-control/wiki/Troubleshooting#warn-skip-generating-metric-sample-for-broker-because-all-broker-metrics-are-missing) you can find more info about this warning. If you don't share your Strimzi configuration, we won't be able to help. You can use [this tools](https://github.com/strimzi/strimzi-kafka-operator/blob/main/tools/report.sh) to collect configurations and logs from your cluster.

**Comment by nsettipa** [comment]
Pls find enclosed configuration and cruise-control logs where warnings are reported for broker pods, metrics missing.
[cruise_Control_warnings.logs.txt](https://github.com/user-attachments/files/17108202/cruise_Control_warnings.logs.txt)
[Uploading cruise_Control_warnings.logs.txt…]()

**Comment by nsettipa** [comment]
Configuration:
apiVersion: mbkf.data.traffic.com/v1
kind: MessageBus
metadata:
  annotations:
    meta.helm.sh/release-name: traf-bss-msg-kafka-cr
    meta.helm.sh/release-namespace: caf-n134
  creationTimestamp: "2024-04-26T15:42:42Z"
  generation: 11
  labels:
    app.kubernetes.io/managed-by: Helm
    appGroup: msg1
    appGroupName: msg1
    traf-data-distributed-coordinator-ed-access: "true"
    traf-sec-key-management-access: "true"
    traf-sec-sip-tls-access: "true"
    msg-release: traf-bss-msg-kafka-cr
  name: traf-bss-msg-msg1
  namespace: caf-n134
  resourceVersion: "1648387561"
  uid: bb655513-1b4b-4a79-b6e5-cd5a252aee51
spec:
  affinity:
    podAntiAffinity: soft
    topologyKey: kubernetes.io/hostname
  authorization:
    enabled: false
    superUsers:
    - CN=traf-bss-msg-kafka
  clusterConfig:
    rack:
      topologyKey: topology.kubernetes.io/zone
    settings:
      default.replication.factor: 3
      leader.imbalance.per.broker.percentage: 0
      log.retention.hours: 1
      log.segment.bytes: "10485760"
      max.partition.fetch.bytes: "104857600"
      message.max.bytes: "104857600"
      min.insync.replicas: 2
      num.partitions: 4
      num.replica.fetchers: 2
      offsets.topic.replication.factor: 3
      replica.fetch.max.bytes: "33554432"
      replica.fetch.min.bytes: "1024"
      socket.receive.buffer.bytes: "33554432"
      socket.request.max.bytes: "104857600"
      socket.send.buffer.bytes: "33554432"
  cruiseControl:
    autoApprovalInterval: 15
    config:
      anomaly.detection.goals: |
        com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.MinTopicLeadersPerBrokerGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.TopicReplicaDistributionGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaDistributionGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.CpuUsageDistributionGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.LeaderReplicaDistributionGoal
      cpu.balance.threshold: 1.1
      default.goals: |
        com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.MinTopicLeadersPerBrokerGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.TopicReplicaDistributionGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaDistributionGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.CpuUsageDistributionGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.LeaderReplicaDistributionGoal
      failed.brokers.file.path: /tmp/failedBrokers.txt
      hard.goals: |
        com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal
      sample.store.topic.replication.factor: 3
      topic.replica.count.balance.threshold: 1
    enableAutomation: false
    enabled: true
  externalExposure:
    configuration:
      advertisedHostPort: 443
      annotations: {}
      bootstrapConfig: {}
      domainName: n134-traf-pool-1.sc.seli.gic.trafsson.se
      labels: {}
      perBrokerConfig: []
    enabled: true
    exposureType: cluster-ip
    port: 9094
    tls:
      customTLSSecrets:
        serverCertificate:
          secretName: traf-bss-msg-kafka-external-server-cert-secret
        trustedCABundle:
          secretName: traf-bss-msg-kafka-external-client-ca-cert
  instanceVersions:
  - 3.5.1
  log:
    level: error
    streamingMethod: indirect
  metrics:
    enabled: true
    jmxFilterConfigMapName: traf-bss-msg-kafka-jmx-metrics-filter-v3
  networkPolicy:
    enabled: false
  nodeSelector:
    traf-bss-app: caf
    traf-bss-app-msg: msg
  podDisruptionBudget:
    minAvailable: 51%
  replicas: 3
  resources:
    flavor: default
    persistentStorage:
      persistentVolumeClaim:
        size: 1000Gi
        storageClassName: caf-pool1
  suspend: false
  tls:
    enforced: required
    port: 9093
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 30
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 30
  topicManagement:
    enabled: false
  topologySpreadConstraints:
  - maxSkew: 1
    minDomains: 3
    topologyKey: topology.kubernetes.io/zone
    whenUnsatisfiable: DoNotSchedule
  - maxSkew: 1
    topologyKey: kubernetes.io/hostname
    whenUnsatisfiable: ScheduleAnyway
  zookeeper:
    affinity:
      podAntiAffinity: hard
      topologyKey: kubernetes.io/hostname
    nodeSelector:
      traf-bss-app: caf
      traf-bss-app-msg: msg
    podDisruptionBudget:
      minAvailable: 51%
    replicas: 3
    resources:
      flavor: default
      persistentStorage:
        persistentVolumeClaim:
          size: 5Gi
          storageClassName: caf-pool1
    tolerations:
    - effect: NoExecute
      key: node.kubernetes.io/not-ready
      operator: Exists
      tolerationSeconds: 30
    - effect: NoExecute
      key: node.kubernetes.io/unreachable
      operator: Exists
      tolerationSeconds: 30
    topologySpreadConstraints:
    - maxSkew: 1
      minDomains: 3
      topologyKey: topology.kubernetes.io/zone
      whenUnsatisfiable: DoNotSchedule
    - maxSkew: 1
      topologyKey: kubernetes.io/hostname
      whenUnsatisfiable: ScheduleAnyway
status:
  clusterId: 7e9ZcLQwT36vZoSXSiBKpQ
  conditions:
  - lastTransitionTime: "2024-09-11T14:10:10.427240780Z"
    status: "True"
    type: Ready
  listeners:
  - addresses:
    - host: traf-bss-msg-msg1-kafka-bootstrap.caf-n134.svc
      port: 9093
    bootstrapServers: traf-bss-msg-msg1-kafka-bootstrap.caf-n134.svc:9093

**Comment by nsettipa** [comment]
Hi,

Is there possibility to provide the solution or recommends. I downloaded
cruise-control and found that below code, where this warning reported.

private static boolean skipBuildingBrokerMetricSample(BrokerLoad
brokerLoad, int brokerId) {
    if (brokerLoad == null) {
      LOG.warn("{}broker {} because all broker metrics are missing.",
SKIP_BUILDING_SAMPLE_PREFIX, brokerId);
      return true;

when workLoad is null, this warning reported. when this warning will be
reported. what is the interpretation of brokerLoad null.

This is found in cruisecontrol/monitor/sampling/SamplingUtils.jav

Thank you.
Rgds,
Narasimham

**Comment by nsettipa** [comment]
Any update on this ticket. Could you pls confirm whether it can be fixed on
cruisecontrol side or any specific broker side changes needed

Broker side kakfa cruise control reporter jar exists.

**Comment by nsettipa** [comment]
Any update on the analysis. Any correction would be expected for this issue. Please let me know. Thanks a lot.

**Reply by scholzj** [reply]
TBH, I'm not sure what exactly you expect here. Nothing that you shared is properly formatted and is there fore unreadable. More over, Strimzi just orchestrates Kafka and Cruise Control. It might not be able to give you an answer for low-level questions about them. You might need to ask about it in the respective communities of given projects.

**Comment by nsettipa** [comment]
The cruise-control log shows below warning: 
if (brokerLoad == null) {
      LOG.warn("{}broker {} because all broker metrics are missing.", SKIP_BUILDING_SAMPLE_PREFIX, brokerId);
      return true;

In the function: private static boolean skipBuildingBrokerMetricSample

Looks like brokerLoad is null, when this could happen. Is it that, broker metrics not received. we ensured metrics reporting jar is running on all kafka broker pods.

**Reply by scholzj** [reply]
Ok - so this kind of question is probably something to ask in the Cruise Control community. Please ask there. I'm not sure if anyone in Strimzi has this level of deep source code knowledge of Cruise Control.

**Comment by nsettipa** [comment]
Could you pls provide community details. I will raise there. Thank you.

**Reply by scholzj** [reply]
The Cruise Control project is on GitHub. Not sure if the use GitHub discussions, issues, if they have some Slack channel etc. You would have to check it out.

**Comment by nsettipa** [comment]
Hi,

I was unable to create ticket/issue/clarifications on cruise-control. could
you please help regarding this?.

Thanks in advance.

Rgds,
Narasimham

**Reply by ppatierno** [reply]
What was so hard to go here https://github.com/linkedin/cruise-control/issues and click on the "New Issue" button?
